### PR TITLE
mm: memory allocations return valid pointer when request 0 size

### PR DESCRIPTION
**Pull request by pkarashchenko**
## Summary
This change introduce 2 items:
1. If the size of the space requested is 0, the behavior is implementation-defined:
   either a null pointer shall be returned, or the behavior shall be as if the size
   were some non-zero value, except that the behavior is undefined if the returned
   pointer is used to access an object.

   Change the behavior to be similar to Linux and Android and allocates an object
   of a minimum size instead of returning null pointer.

   https://pubs.opengroup.org/onlinepubs/9699919799/functions/malloc.html
   https://pubs.opengroup.org/onlinepubs/9699919799/functions/calloc.html
   https://pubs.opengroup.org/onlinepubs/9699919799/functions/realloc.html

2. The description of realloc() has been modified from previous versions of this
   standard to align with the ISO/IEC 9899:1999 standard. Previous versions explicitly
   permitted a call to realloc (p, 0) to free the space pointed to by p and return
   a null pointer. While this behavior could be interpreted as permitted by this
   version of the standard, the C language committee have indicated that this
   interpretation is incorrect. Applications should assume that if realloc() returns
   a null pointer, the space pointed to by p has not been freed. Since this could lead
   to double-frees, implementations should also set errno if a null pointer actually
   indicates a failure, and applications should only free the space if errno was changed.

   Do not free memory of zero-length reallocation is requested

   https://pubs.opengroup.org/onlinepubs/9699919799/functions/realloc.html

## Impact
Change the behaviour of zero-length allocation across the system.
Improve POSIX compliance of `realloc`

## Testing
Pass CI. Tests in progress

### CODE REVIEW

**Review by patacongo — CHANGES_REQUESTED** [review]
These memory management function changes would set the `errno`.  So when these memory management functions are called from the OS, the `errno` may be trashed.  Kernel logic must never modifiy the `errno` value per Issue #1027.

In order to make this change, the memory management subsystem architecture will need to be modified to permit OS access to the allocators without modifying the `errno`.  For example, perhaps there is an` nx_realloc()`, for example, that is called from `mm_realloc()` like:

      mem = nx_realloc(USR_HEAP, alignment, size);
      if (!mem)
        {
          set_errno(ENOMEM);
        }

**Comment by pkarashchenko** [comment]
> These memory management function changes would set the `errno`. So when these memory management functions are called from the OS, the `errno` may be trashed. Kernel logic must never modifiy the `errno` value per Issue #1027.
> 
> In order to make this change, the memory management subsystem architecture will need to be modified to permit OS access to the allocators without modifying the `errno`. For example, perhaps there is an` nx_realloc()`, for example, that is called from `mm_realloc()` like:
> 
> ```
>   mem = nx_realloc(USR_HEAP, alignment, size);
>   if (!mem)
>     {
>       set_errno(ENOMEM);
>     }
> ```

I will revert errno changes and keep only zero-allocation logic, so `errno` compliance will be handled in a separate change

**Comment by xiaoxiang781216** [comment]
> > These memory management function changes would set the `errno`. So when these memory management functions are called from the OS, the `errno` may be trashed. Kernel logic must never modifiy the `errno` value per Issue #1027.
> > In order to make this change, the memory management subsystem architecture will need to be modified to permit OS access to the allocators without modifying the `errno`. For example, perhaps there is an` nx_realloc()`, for example, that is called from `mm_realloc()` like:
> > ```
> >   mem = nx_realloc(USR_HEAP, alignment, size);
> >   if (!mem)
> >     {
> >       set_errno(ENOMEM);
> >     }
> > ```
> 
> I will revert errno changes and keep only zero-allocation logic, so `errno` compliance will be handled in a separate change

I think set_errno could move into malloc/calloc.

**Comment by pkarashchenko** [comment]
> > > These memory management function changes would set the `errno`. So when these memory management functions are called from the OS, the `errno` may be trashed. Kernel logic must never modifiy the `errno` value per Issue #1027.
> > > In order to make this change, the memory management subsystem architecture will need to be modified to permit OS access to the allocators without modifying the `errno`. For example, perhaps there is an` nx_realloc()`, for example, that is called from `mm_realloc()` like:
> > > ```
> > >   mem = nx_realloc(USR_HEAP, alignment, size);
> > >   if (!mem)
> > >     {
> > >       set_errno(ENOMEM);
> > >     }
> > > ```
> > 
> > 
> > I will revert errno changes and keep only zero-allocation logic, so `errno` compliance will be handled in a separate change
> 
> I think set_errno could move into malloc/calloc.

I places set_errno into malloc/calloc only, but I think @patacongo is talking about FLAT build where kernel allocations are remapped to `malloc/calloc`.

**Comment by xiaoxiang781216** [comment]
This can be fixed by stopping map kmm_malloc to malloc directly in FLAT too.

File: mm/umm_heap/umm_realloc.c [hunk 1]
```diff
@@ -83,7 +77,7 @@ FAR void *realloc(FAR void *oldmem, size_t size)
       mem = mm_realloc(USR_HEAP, oldmem, size);
       if (!mem)
         {
-          brkaddr = sbrk(size);
+          brkaddr = sbrk(size < 1 ? 1 : size);
```
Comment (patacongo) [inline mm/umm_heap/umm_realloc.c:80]: The actual size needed is actually larger than one when size is zero:  It is the minimum chunk size.

File: mm/tlsf/mm_tlsf.c [hunk 2]
```diff
@@ -1124,15 +1130,20 @@ FAR void *mm_realloc(FAR struct mm_heap_s *heap, FAR void *oldmem,
...
-  if (size == 0)
+  /* If size is zero, reallocate to the minim size object, so
+   * the memory pointed by oldmem is freed
+   */
+
+  if (size < 1)
     {
-      mm_free(heap, oldmem);
-      return NULL;
+      size = 1;
```
Comment (patacongo) [inline mm/tlsf/mm_tlsf.c:1146]: Passing size=0 and size=1 to mm_malloc have the same effect. A minimum chunk size is allocated. So I think the above can be removed.

File: mm/tlsf/mm_tlsf.c [hunk 3]
```diff
@@ -1005,6 +1004,13 @@ FAR void *mm_malloc(FAR struct mm_heap_s *heap, size_t size)
 {
   FAR void *ret;
 
+  /* In case of zero-length allocations allocate the minimum size object */
+
+  if (size < 1)
+    {
+      size = 1;
+    }
+
```
Comment (patacongo) [inline mm/tlsf/mm_tlsf.c:1013]: Passing size=0 and size=1 to mm_malloc have the same effect. A minimum chunk size is allocated. So I think the above can be removed.

File: mm/umm_heap/umm_realloc.c (hunk 1 above)
Comment (patacongo) [inline mm/umm_heap/umm_realloc.c:80]:
```suggestion
          brkaddr = sbrk(size < MM_MIN_CHUNK ? MM_MIN_CHUNK : size);
```

**Comment by pkarashchenko** [comment]
> This can be fixed by stopping map kmm_malloc to malloc directly in FLAT too. From errno perspective, kmm_malloc isn't equals to malloc anyway.

Anyway we need a kind of a wrapper here. So let's move this part to another PR.

File: mm/umm_heap/umm_realloc.c (hunk 1 above)
Comment (pkarashchenko) [reply mm/umm_heap/umm_realloc.c:80]: Same answer as above. `MM_MIN_CHUNK` is not valid for `tlsf_realloc`

Comment (pkarashchenko) [reply mm/umm_heap/umm_realloc.c:80]:
```
TLSF_API void* tlsf_realloc(tlsf_t tlsf, void* ptr, size_t size)
{
	control_t* control = tlsf_cast(control_t*, tlsf);
	void* p = 0;

	/* Zero-size requests are treated as free. */
	if (ptr && size == 0)
	{
		tlsf_free(tlsf, ptr);
	}
```
so `tlsf_realloc` is not compliant with the latest POSIX spec

**Review by patacongo — CHANGES_REQUESTED** [review]
See comments

File: mm/umm_heap/umm_realloc.c (hunk 1 above)
Comment (pkarashchenko) [reply mm/umm_heap/umm_realloc.c:80]: Will handle according to https://pubs.opengroup.org/onlinepubs/7908799/xsh/brk.html

File: mm/umm_heap/umm_memalign.c [hunk 4]
```diff
@@ -75,7 +75,7 @@ FAR void *memalign(size_t alignment, size_t size)
       mem = mm_memalign(USR_HEAP, alignment, size);
       if (!mem)
         {
-          brkaddr = sbrk(size);
+          brkaddr = sbrk(size < 1 ? 1 : size);
```
[the remark this answers is not in the document]
Comment (pkarashchenko) [reply mm/umm_heap/umm_memalign.c:78]: I will take a look if there is an option to get a minimal chunk size from `mm_` layer. Anyway this is valid only if the min chunk is greater than the page size.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by pkarashchenko** [comment]
Now it is ready

**Comment by pkarashchenko** [comment]
@patacongo @xiaoxiang781216 should we remove `set_errno(errcode);` from `FAR void *sbrk(intptr_t incr)` implementation for now until we fix the https://github.com/apache/nuttx/issues/1027 and add corresponding wrappers of allocation API in case of FLAT mode?

**Review by xiaoxiang781216 — APPROVED** [review]

File: mm/umm_heap/umm_realloc.c (hunk 1 above)
Comment (patacongo) [reply mm/umm_heap/umm_realloc.c:80]: > Same answer as above. `MM_MIN_CHUNK` is not valid for `tlsf_realloc`

But there still could be an issue here.  Not just for size < 1 for for all sizes.  The sizes don't include quantization overhead, or the overhead from the memory header.  So the amount of memory requested is smaller than that which is actually needed.  If nothing else, that does leave me a little uneasy.

If tlsf_realloc and mm_realloc are different, that would not change the issue, that would just make any solution that might be needed more difficult (and I'm not even certain that one is needed).  I'll open an issue so we don't forget, but let's decouple this from the change at hand in the interest of making progress.

Issue #9155

Comment (pkarashchenko) [reply mm/umm_heap/umm_realloc.c:80]: Agreed. Thank you!

**Comment by patacongo** [comment]
There are two approvals and no open discussions.  I will go ahead and merge without waiting for @hartmannathan

**Comment by hartmannathan** [comment]
> There are two approvals and no open discussions.  I will go ahead and merge without waiting for @hartmannathan 

Looks good to me. Thanks.

**Pull request merged**